sage_modabvar/abvar.py

```diff
@@ -2129,6 +2130,38 @@ def _ambient_hecke_matrix_on_modular_symbols(self, n):
         self.__ambient_hecke_matrix_on_modular_symbols[n] = T
         return T
 
+    def number_of_rational_points(self):
+        """
+        Return the number of rational points of this modular abelian variety.
+
+        It is not always possible to compute the order of the torsion
+        subgroup. The BSD conjecture is assumed to compute the algebraic rank.
+
+        OUTPUT: a positive integer or infinity
+
+        EXAMPLES::
+
+        sage: from sage_modabvar import J0, J1
+        sage: J0(23).number_of_rational_points()
+        11
+        sage: J0(29).number_of_rational_points()
+        7
+        sage: J0(37).number_of_rational_points()
+        +Infinity
+        """
+
+        # The rank of a simple is positive if the lratio is zero.
+        positive_rank_simple = lambda simple: simple.lseries().lratio() == 0
+
+        # The rank is positive if at least one simple has positive rank
+        positive_rank = any(positive_rank_simple(simple)
+                for simple in self.decomposition())
+
+        if positive_rank:
+            return infinity
+        else:
+            return self.rational_torsion_subgroup().order()
```

Comment:
include an example in the doctests whether this fails, e.g., where `.rational_torsion_subgroup().order()` doesn't work.  There should be a lot, since there's no algorithm to compute `.rational_torsion_subgroup().order()` in general.